src/python/pants/backend/experimental/jvm/goals/coursier.py

```diff
@@ -0,0 +1,151 @@
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os.path
+from dataclasses import dataclass
+
+from pants.backend.experimental.jvm.resolve.coursier_fetch import (
+    CoursierResolvedLockfile,
+    MavenRequirements,
+)
+from pants.backend.experimental.jvm.target_types import JvmLockfileSources, MavenRequirementsField
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.console import Console
+from pants.engine.fs import (
+    EMPTY_DIGEST,
+    CreateDigest,
+    Digest,
+    DigestContents,
+    FileContent,
+    MergeDigests,
+    Snapshot,
+    Workspace,
+)
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.rules import Get, MultiGet, collect_rules, goal_rule, rule
+from pants.engine.target import Sources, Target, Targets
+
+
+class CoursierResolveSubsystem(GoalSubsystem):
+    name = "coursier-resolve"
+    help = "Generate a lockfile by resolving Maven dependencies."
+
+
+class CoursierResolve(Goal):
+    subsystem_cls = CoursierResolveSubsystem
+
+
+@dataclass(frozen=True)
+class CoursierGenerateLockfileRequest:
+    target: Target
+
+
+@dataclass(frozen=True)
+class CoursierGenerateLockfileResult:
+    digest: Digest
+
+
+@rule
+async def coursier_generate_lockfile(
+    request: CoursierGenerateLockfileRequest,
+) -> CoursierGenerateLockfileResult:
+    """Regenerate a coursier_lockfile target's lockfile from its Maven requirements.
+
+    This allows the user to manually regenerate their lockfile.  This is done for a few reasons: to
+    generate the lockfile for the first time, to regenerate it because the input Maven requirements
+    have changed, or to regenerate it to check if the resolve has changed (e.g. due to newer
+    versions of dependencies being published).
+    """
+    resolved_lockfile = await Get(
+        CoursierResolvedLockfile,
+        MavenRequirements,
+        MavenRequirements.create_from_maven_coordinates_fields(
+            fields=(request.target[MavenRequirementsField],),
+        ),
+    )
+    resolved_lockfile_json: bytes = resolved_lockfile.to_json()
+
+    lockfile_sources = await Get(
+        SourceFiles,
+        SourceFilesRequest(
+            [request.target.get(Sources)],
+            for_sources_types=[JvmLockfileSources],
+            enable_codegen=False,
+        ),
+    )
+    if not lockfile_sources.files:
+        # The user defined the target and resolved it, but hasn't created a lockfile yet.
+        # For convenience, create the initial lockfile for them in the default path.
+        return CoursierGenerateLockfileResult(
+            digest=await Get(
+                Digest,
+                CreateDigest(
+                    (
+                        FileContent(
+                            path=os.path.join(
+                                request.target.address.spec_path, "coursier_resolve.lockfile"
```

Comment:
pathlib is probably better, no ? @Eric-Arellano

Reply:
I used to think so, but then found pathlib has worse performance so no longer really ask for people to make changes, especially when there isn't much to gain like `os.path.join`. Only time I ask for pathlib is replacing:

```
with open(fp) as f:
   content = f.read()
```

with:

```
content = Path(fp).read_text()
```